blog/_posts/2025-04-04-Optimize-AKS-Traffic-with-externalTrafficPolicy=Local.md

```diff
@@ -0,0 +1,157 @@
+
+In this blog, we delve into the intricacies of Kubernetes’ `externalTrafficPolicy=Local` setting and explore strategies to gracefully handle pod shutdowns, rolling updates, and pod distribution. By following these best practices, you can enhance the resilience and reliability of your services while optimizing resource utilization across your AKS clusters.
+
+## `externalTrafficPolicy`: Cluster vs. Local
+
+**What does `externalTrafficPolicy` do?** In short, it influences how a cloud load balancer directs incoming traffic to Kubernetes nodes and how that traffic may hop between nodes to reach pods:
+
+- **Cluster Mode (default) (`externalTrafficPolicy=Cluster`)**:
+    This is default mode for Kubernetes Services and provides a simplified approach to traffic distribution across all nodes in the cluster. While it does not preserve the client source IP, it offers several benefits for certain workloads:
+
+  - **Simplified Traffic Distribution**: Ensures even distribution of incoming traffic across all nodes in the cluster, regardless of pod placement.
+  - **Reduced Risk of Node Overload**: Balances traffic across all nodes, minimizing the chance of overloading specific nodes.
+  - **Ease of Configuration**: Requires no additional configuration for health checks or pod distribution.
+  - **Resilience During Pod Failures**: Traffic is forwarded to healthy pods on other nodes, ensuring service availability even if some pods are unavailable.
+
+    However, because this mode distributes traffic to all nodes regardless of them hosting the service pod, it can add latency and more failure modes due to extra hops between nodes (forwarded by `kube-proxy`). This is especially detrimental during events like node upgrades since if a node is taken down, 1/N (N stands for node count) traffic could experience disruptions. So, if you update your nodes weekly, this would disrupt your connections on a weekly basis.
+
+- **Local Mode (`externalTrafficPolicy=Local`)**:
+    In this mode, external traffic is directed only to nodes that host healthy pods for the specified service. The traffic is routed exclusively to the pods residing on the same node where it is received (avoiding cross-node hops). This approach ensures:
+  - **Localized Impact of Node Downtime**: During operations like upgrades, reboots and unexpected failures on a node, a portion of external traffic is only affected if the downed node is running a service pod. The traffic to the pods on different nodes remains unaffected.
+  - **Preservation of Client Source IP**: The client’s original IP is maintained, which is crucial for security, logging, and analytics.
+  - **Targeted Health Checking**: Local mode uses a dedicated `healthCheckNodePort`, ensuring that external load balancers send traffic only to nodes with healthy pods.
```

Comment:
Targeted Health Checking feels more like an implementation detail than a benefit.

Reply:
I have removed it from this benefit section.